README.md
## What is this?

A virtual representation of the clock used in Klaushiere, a fictional world appears in Takahiko Mukoyama's novel called 童話物語 (Dowa Monogatari, lit. Fairy tale Story).

## Key features

* Wheels move in real-time
* Inactive wheels will be dimmed automatically based on time of the day
* Besides of the clock itself, It shows three digits time in Klausherian time system
* It can be used like native phone app by adding it to the home screen

You can try it out right now from here:
https://klaushierian-clock.manabumiwa.jp

## Available Scripts

In the project directory, you can run:

### `npm start`

Runs the app in the development mode.\
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.\
You will also see any lint errors in the console.

### `npm test`

Launches the test runner in the interactive watch mode.\
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `npm run build`

Builds the app for production to the `build` folder.\
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.\
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `npm run eject`

**Note: this is a one-way operation. Once you `eject`, you can’t go back!**

If you aren’t satisfied with the build tool and configuration choices, you can `eject` at any time. This command will remove the single build dependency from your project.

Instead, it will copy all the configuration files and the transitive dependencies (webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you’re on your own.

You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.


CHANGELOG.md
# CHANGELOG

## Version 1.0.0

* Initial release

## Version 1.0.1

* Tweak responsive behavior


src/klaushiere-time.spec.ts
import { KlaushiereTime, Periods } from './klaushiere-time'

describe('KlaushiereTime class', () => {
  it('should show each periods correctly', () => {
    const time = new Date()
    let subject: KlaushiereTime

    type TestTableColumn = {
      hours: number
      expected: Periods
    }

    const testTable: TestTableColumn[] = [
      { hours: 0, expected: 'pre' },
      { hours: 9, expected: 'pre' },
      { hours: 10, expected: 'rep' },
      { hours: 19, expected: 'rep' },
      { hours: 20, expected: 'dapre' },
      { hours: 23, expected: 'dapre' },
    ]

    testTable.forEach(c => {
      time.setHours(c.hours)
      subject = new KlaushiereTime(time)
      expect(subject.period).toBe(c.expected)  
    })
  })

  it('should show correct minutes', () => {
    const time = new Date()
    let subject: KlaushiereTime

    type TestTableColumn = {
      hours: number
      minutes: number
      seconds: number
      expected: number
    }

    const testTable: TestTableColumn[] = [
      { hours: 0, minutes: 0, seconds: 0, expected: 0 },
      { hours: 7, minutes: 0, seconds: 0, expected: 700 },
      { hours: 9, minutes: 59, seconds: 59, expected: 999 },
      { hours: 10, minutes: 0, seconds: 0, expected: 0 },
      { hours: 19, minutes: 59, seconds: 59, expected: 999 },
      { hours: 20, minutes: 0, seconds: 0, expected: 0 },
      { hours: 23, minutes: 59, seconds: 59, expected: 399 },
    ]

    testTable.forEach(c => {
      time.setHours(c.hours, c.minutes, c.seconds)
      subject = new KlaushiereTime(time)
      expect(subject.minutes).toBe(c.expected)  
    })
  })
})


src/klaushierian-time.test.ts
import { KlaushierianTime, Periods } from './klaushierian-time'

describe('KlaushierianTime class', () => {
  let subject: KlaushierianTime
  const date = '1970-01-01T'

  describe('minutes getter', () => {
    it('should return correct Klaushierian minutes', () => {
      type TestCase = {
        time: string
        expected: number
      }

      const testCases: TestCase[] = [
        { time: '00:00:00', expected: 0 },
        { time: '01:00:00', expected: 100 },
        { time: '05:30:00', expected: 550 },
        { time: '09:59:59', expected: 999 },
        { time: '10:00:00', expected: 0 },
        { time: '11:00:00', expected: 100 },
        { time: '15:30:00', expected: 550 },
        { time: '19:59:59', expected: 999 },
        { time: '20:00:00', expected: 0 },
        { time: '21:00:00', expected: 100 },
        { time: '23:59:59', expected: 399 }
      ]
      testCases.forEach(c => {
        subject = new KlaushierianTime(new Date(date + c.time))
        expect(subject.minutes).toBe(c.expected)
      })
    })
  })

  describe('period getter', () => {
    it('should return correct Klaushierian period', () => {
      type TestCases = {
        time: string
        expected: Periods
      }
      
      const testCases: TestCases[] = [
        { time: '00:00:00', expected: 'pre' },
        { time: '01:00:00', expected: 'pre' },
        { time: '05:30:00', expected: 'pre' },
        { time: '09:59:59', expected: 'pre' },
        { time: '10:00:00', expected: 'rep' },
        { time: '11:00:00', expected: 'rep' },
        { time: '15:30:00', expected: 'rep' },
        { time: '19:59:59', expected: 'rep' },
        { time: '20:00:00', expected: 'dapre' },
        { time: '21:00:00', expected: 'dapre' },
        { time: '23:59:59', expected: 'dapre' }
      ]

      testCases.forEach(c => {
        subject = new KlaushierianTime(new Date(date + c.time))
        expect(subject.period).toBe(c.expected)
      })
    })
  })

  describe('toString method', () => {
    it('should return string representation of corresponding Klaushierian time', () => {
      type TestCase = {
        time: string
        expected: string
      }

      const testCases: TestCase[] = [
        { time: '00:00:00', expected: '0 pre' },
        { time: '01:00:00', expected: '100 pre' },
        { time: '05:30:00', expected: '550 pre' },
        { time: '09:59:59', expected: '999 pre' },
        { time: '10:00:00', expected: '0 rep' },
        { time: '11:00:00', expected: '100 rep' },
        { time: '15:30:00', expected: '550 rep' },
        { time: '19:59:59', expected: '999 rep' },
        { time: '20:00:00', expected: '0 dapre' },
        { time: '21:00:00', expected: '100 dapre' },
        { time: '23:59:59', expected: '399 dapre' }
      ]

      testCases.forEach(c => {
        subject = new KlaushierianTime(new Date(date + c.time))
        expect(subject.toString()).toBe(c.expected)
      })
    })
  })

  describe('outerAngle getter', () => {
    it('should return angle of outer wheel', () => {
      type TestCase = {
        time: string
        expected: number
      }
  
      const testCases: TestCase[] = [
        { time: '00:00:00', expected: 360 },
        { time: '03:00:00', expected: 252 },
        { time: '06:00:00', expected: 144 },
        { time: '09:00:00', expected: 36 },
        { time: '09:59:59', expected: 0.009999999999990905 }
      ]  

      testCases.forEach(c => {
        subject = new KlaushierianTime(new Date(date + c.time))
        expect(subject.outerAngle).toBe(c.expected)
      })
    })
  })

  describe('midAngle getter', () => {
    it('should return angle of mid wheel', () => {
      type TestCase = {
        time: string
        expected: number
      }
  
      const testCases: TestCase[] = [
        { time: '00:00:00', expected: 0 },
        { time: '00:15:00', expected: 90 },
        { time: '00:30:00', expected: 180 },
        { time: '00:45:00', expected: 270 },
        { time: '03:00:00', expected: 0 },
        { time: '05:15:00', expected: 90 },
        { time: '07:30:00', expected: 180 },
        { time: '09:45:00', expected: 270 },
        { time: '09:59:59', expected: 359.89999999999964 }
      ]  

      testCases.forEach(c => {
        subject = new KlaushierianTime(new Date(date + c.time))
        expect(subject.middleAngle).toBe(c.expected)
      })
    })
  })

  describe('innerAngle getter', () => {
    it('should return angle of inner wheel', () => {
      type TestCase = {
        time: string
        expected: number
      }
  
      const testCases: TestCase[] = [
        { time: '00:00:00', expected: 0 },
        { time: '09:59:59', expected: 0 },
        { time: '10:00:00', expected: 0 },
        { time: '11:00:00', expected: 0 },
        { time: '19:59:59', expected: 0 },
        { time: '20:00:00', expected: 360 },
        { time: '22:00:00', expected: 180 },
        { time: '23:59:59', expected: 0.025000000000034106 }
      ]  

      testCases.forEach(c => {
        subject = new KlaushierianTime(new Date(date + c.time))
        expect(subject.innerAngle).toBe(c.expected)
      })
    })
  })
})


src/klaushiere-time.ts
export type Periods = 'pre' | 'rep' | 'dapre'

/** Class replesents time system used in Klaushiere */
export class KlaushiereTime {
  private _time: Date

  /**
   * @param [time=new Date()] - A Date object used for initializing a instance
   */
  constructor(time?: Date) {
    this._time = time ?? new Date()
  }

  get minutes(): number {
    const h = this._time.getHours()
    const m = this._time.getMinutes()
    const s = this._time.getSeconds()
    const ms = this._time.getMilliseconds()

    const timeInMs = h * 3600000 + m * 60000 + s * 1000 + ms
    return Math.floor(timeInMs / 36000) % 1000
  }

  get period(): Periods {
    const hours = this._time.getHours()
    if (hours < 10) return 'pre'
    if (hours < 20) return 'rep'
    return 'dapre'
  }

  /** Returns a string representation of a instance */
  toString() {
    return `${this.minutes} ${this.period}`
  }
}


src/klaushierian-time.ts
export type Periods = 'pre' | 'rep' | 'dapre'

/** Class replesents time system used in Klaushiere */
export class KlaushierianTime {
  private _time: Date

  /**
   * @param [time=new Date()] - A Date object used for initializing a instance
   */
  constructor(time?: Date) {
    this._time = time ?? new Date()
  }

  private get _klaushierianTime(): number {
    const h = this._time.getHours()
    const m = this._time.getMinutes()
    const s = this._time.getSeconds()
    const ms = this._time.getMilliseconds()

    const timeInMs = h * 3600000 + m * 60000 + s * 1000 + ms
    return timeInMs / 36000 % 1000
  }

  get minutes(): number {
    return Math.floor(this._klaushierianTime)
  }

  get period(): Periods {
    const hours = this._time.getHours()
    if (hours < 10) return 'pre'
    if (hours < 20) return 'rep'
    return 'dapre'
  }

  get outerAngle(): number {
    return this.period !== 'dapre'
      ? 360 - this._klaushierianTime / (1000 / 360)
      : 0
  }

  get middleAngle(): number {
    return this._klaushierianTime / (100 / 360) % 360
  }

  get innerAngle(): number {
    return this.period === 'dapre'
      ? 360 - this._klaushierianTime / (400 / 360)
      : 0
  }

  /** Returns a string representation of a instance */
  toString() {
    return `${this.minutes} ${this.period}`
  }
}
